New PropertyExpression as utility class

Mwexim opened the pull request:
- Added PropertyExpression
  - It works almost the same as ConditionalExpression and its sole purpose is utility.
  - The patterns are actually created for you, so all the work is off the table.
  - This makes properties very straightforward, like in vanilla Skript, except it is more clear now.

- Currently, I added only small features and maybe with some more input this feature can be extended more.

- Maybe we can change some of the current expressions (like ExprAmount) to properties to showcase this.

If there are any remarks, please let me know! I tested the code locally, so normally it should work fine.

Mwexim committed:
Added property expressions

Added a utility class for properties.

Mwexim committed:
Added property registration

Added the ability to register properties.

Mwexim committed:
More utility

Added a default init method. It can be overridden when needed.

Syst3ms commented:
Alright, this is good, I'm just gonna ask for two changes for consistency with the current API :

1. The methods should be "addPropertyExpression"
2. There should also be a "newPropertyExpression" method

Mwexim committed:
Added requests

- Added the consistency requests and the new method

Mwexim committed:
Added more utility

- Added javadocs for more information
- Added a #getPropertyFunction() method that automatically applies the values
- Fixed an issue that'd cause bugs.
  - I'm essentially creating this with my own package names, and I just saw that these ones were still called after my packages. I renamed them all to the correct ones, but please check if they are correct because this would've caused bugs.

Mwexim commented:
I added some more utility and fixed a critical bug. 
As always, the new methods were tested on my local server and everything worked fine. 
If this gets pushed, we can start replacing some of the existing expressions with properties.

Mwexim committed:
Added last simple change

Mwexim committed:
Made #getPropertyFunction() non-abstract since you don't always want to use it.

Syst3ms commented on PropertyExpression.java:
```diff
@@ -0,0 +1,99 @@
+    @Override
+    public T[] getValues(TriggerContext ctx) {
+        O[] objs = getOwner().getValues(ctx);
+        if (objs == null) return (T[]) new Object[0];
```
```suggestion
        if (objs.length == 0) return (T[]) new Object[0];
```
The specification of getValues() says it shouldn't ever return null, rather an empty array. It should never, ever return null, and it isn't skript-parser's responsibility to ensure it doesn't.

Syst3ms commented on PropertyExpression.java:
```diff
@@ -0,0 +1,99 @@
+    public T[] getValues(TriggerContext ctx) {
+        O[] objs = getOwner().getValues(ctx);
+        if (objs == null) return (T[]) new Object[0];
+        if (getPropertyFunction() == null) return (T[]) new Object[0];
```
It doesn't really make sense for this check to just return null (well, the equivalent thereof). If someone overrides getPropertyFunction() in any reasonable fashion, the result would be non-null. And if they don't, it would be nonsensical not to override getValues() instead.

So I think this should instead throw an UnsupportedOperationException, and then it should be mentioned in the specification that if getPropertyFunction() isn't overridden, then getValues() must be.

Syst3ms commented on PropertyExpression.java:
```diff
@@ -0,0 +1,99 @@
+    @Override
+    public boolean init(Expression<?>[] expressions, int matchedPattern, ParseContext parseContext) {
+        setOwner((Expression<O>) expressions[0]);
+        return getOwner() != null;
```
```suggestion
        return true;
```
`expressions[0]` (and getOwner()) cannot possibly be null here (and if they are, we have much bigger problems than this check could ever solve).

Syst3ms commented on PropertyExpression.java:
```diff
@@ -0,0 +1,99 @@
+    }
+
+    /**
+     * This {@code init()} method overrides the default method and does the initialization by itself.
```
```suggestion
     * This default {@code init()} implementation automatically properly sets the owner of this property,
     * which can be accessed using {@link #getOwner()}. If this implementation is overridden for one reason
     * or another, it must call {@link #setOwner(Expression<O>)} properly.
```
I think this specification would be more appropriate.

Syst3ms reviewed (changes requested):
Very good work so far, I'd just have a couple more changes to make before merging.

Mwexim committed:
Apply suggestions from code review

First batch, one suggestion is still open.

Co-authored-by: Syst3ms <arthmm14@gmail.com>

Mwexim committed:
Last request
